### PR TITLE
Move the await outside of the for loop in the Routing::calculate func…

### PR DESCRIPTION
**Pull request by greenscientist**
…tion

The await in the for loop made the call to each RoutingMode sequential. The the promise generation and the promise.all after the loop, we can make all the API call in parallel. Those are mostly fast OSRM calls which are quite fast, but we can still have a small gain.

We moved the core of the calculate function into a private calculateRoutingMode function. The later handle a single mode, and the former handle the loop, the awaiting and the result handling.

Closes: #1028

### CODE REVIEW

**Comment by greenscientist** [comment]
Further benchmark notes: For a route from the north shore of MTL to about St-Jean, it took previously about 0.3s on my computer and with the change about 0.05s.

**Comment by greenscientist** [comment]
Done with the help of Meta.ai

**Review by kaligrafy — APPROVED** [review]

File: packages/chaire-lib-backend/src/services/routing/Routing.ts [hunk 1]
```diff
@@ -121,60 +121,77 @@ export class Routing {
         const { isCancelled } = options;
 
         // ** backend
-        const routingResult: TransitOrRouteCalculatorResult[] = [];
+        const routingPromises: Promise<TransitOrRouteCalculatorResult>[] = [];
```
Comment (GabrielBruno24) [inline packages/chaire-lib-backend/src/services/routing/Routing.ts:124]: Should be routingResultPromises

File: packages/chaire-lib-backend/src/services/routing/Routing.ts [hunk 2]
```diff
@@ -121,60 +121,77 @@ export class Routing {
...
-                            `cannot calculate routing for mode ${routingMode}:  ${error}`,
-                            'TRCalculatorError',
-                            { text: 'transit:transitRouting:errors:ErrorForMode', params: { mode: routingMode } }
-                        )
-                        : error
-                });
-            }
+        // Cancel further processing if the request was cancelled
+        if (isCancelled && isCancelled()) {
+            throw 'Cancelled';
```
Comment (GabrielBruno24) [inline packages/chaire-lib-backend/src/services/routing/Routing.ts:136]: Do we need this both in the loop and after?

File: packages/chaire-lib-backend/src/services/routing/Routing.ts (hunk 1 above)
Comment (GabrielBruno24) [inline packages/chaire-lib-backend/src/services/routing/Routing.ts:124]: Should be routingResultPromises

File: packages/chaire-lib-backend/src/services/routing/Routing.ts (hunk 2 above)
Comment (GabrielBruno24) [inline packages/chaire-lib-backend/src/services/routing/Routing.ts:136]: Do we need this both in the loop and after?

Comment (greenscientist) [reply packages/chaire-lib-backend/src/services/routing/Routing.ts:136]: The one in the loop check cancellation at the beginning, the one outside check after we added all the calls. The idea with cancellation is that you need to check often to interrupt the processing. 
It is really needed, I don't know, we should check in which context it's really used.

File: packages/chaire-lib-backend/src/services/routing/Routing.ts (hunk 1 above)
Comment (greenscientist) [reply packages/chaire-lib-backend/src/services/routing/Routing.ts:124]: Could be. Probably not worth it to change it. The function is small enough that it's easy to see what it is.

**Review by samuel-duhaime — APPROVED** [review]

**Review by GabrielBruno24 — APPROVED** [review]

**Pull request merged**